src/vm/eventpipe.cpp

```diff
@@ -342,6 +343,25 @@ void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallback
     // Disable pSession tracing.
     s_config.Disable(*pSession, pEventPipeProviderCallbackDataQueue);
 
+    // At this point, we should not be writing events to this session anymore
+    // This is a good time to remove the session from the array.
+    _ASSERTE(s_pSessions[pSession->GetIndex()] == pSession);
+
+    // Remove the session from the array, and mask.
+    s_pSessions[pSession->GetIndex()].Store(nullptr);
+
+    // Between here and a few later where we set s_pZombieSession back to nullptr
+    // The session is in zombie state, it is removed from s_pSessions so that 
+    // The YIELD_WHILE in EventPipeBufferManager will always terminate, and yet
+    // EventPipeConfiguration::ComputeKeywordAndLevel() and
+    // EventPipeConfiguration::ComputeEventEnabledMask() will still be able to
+    // compute the right bitmask for filtering events.
+    //
+    // The field is always accessed (read/write) under EventPipe::GetLock() and 
+    // therefore it is impossible for any threads other than the current thread
+    //  to observe s_pZombieSession != null.
+    s_pZombieSession = pSession;
```

Comment:
This works now, but the plan is to stop treating Disable() as a big slow atomic operation which means Rundown could happen in parallel. One zombie session would no longer be sufficient.

One alternative would be using a bitmask to track which sessions are accepting events and update WriteInternal to check the bitmask instead of s_pSessions.

Reply:
Is it the goal to make two rundown runs in parallel? Rundown itself takes the CodeManager lock, meaning two rundowns is likely to be sequential by themselves.

Reply:
I don't think parallelizing two rundowns is critical, my bigger concern was not blocking all the other fast operations such as starting a session, modifying a sessions, or stopping a session that is not doing rundown (Sung is working on that). Given two solutions of roughly equal complexity though, I would favor the one that supports parallel rundown. I think of the CodeManager lock as an unfortunate implementation detail that may someday be remedied.